web-api/src/storage/postgres.rs

```diff
@@ -453,6 +463,88 @@ impl Database {
             .map_err(Into::into)
     }
 
+    #[cfg(feature = "ET-4089")]
+    async fn set_candidates(
+        &self,
+        ids: impl IntoIterator<IntoIter = impl Clone + ExactSizeIterator<Item = &DocumentId>>,
+    ) -> Result<(Vec<DocumentId>, Vec<IngestedDocument>, Warning<DocumentId>), Error> {
+        let mut tx = self.pool.begin().await?;
+
+        let mut ingestable = ids.into_iter().collect::<HashSet<_>>();
+        let unchanged = sqlx::query_as::<_, DocumentId>(
+            "UPDATE document SET
+            is_candidate = FALSE
+            RETURNING document_id;",
```

Comment:
:wrench:  If you want to use it with `RETURNING` you still need to add a where clause, this will always return all documents.

```suggestion
            "UPDATE document SET
            is_candidate = FALSE
            WHERE is_candidate
            RETURNING document_id;",
```

Reply:
:wrench: Also that now would return _all_ document which previously had been candidates  independent of weather or not they still are candidates.

So you still would need to set diff that with the set of new candidates.

Reply:
no, this only returns ids where the flag actually changed, as the [pg docs state](https://www.postgresql.org/docs/current/sql-update.html):
> The optional RETURNING clause causes UPDATE to compute and return value(s) based on each row actually updated.

Reply:
> So you still would need to set diff that with the set of new candidates.

this already happens in the filtermap a few lines below the query.

Reply:
> The optional RETURNING clause causes UPDATE to compute and return value(s) based on each row actually updated.

You update all row, some already having the same value you update it to and postgres internally being able to optimize the writes away due to an index doesn't change the semantics of the query/outcome.

Reply:
i see, i misunderstood the semantics in the docs, my bad, i'll update it.